Add -T install/usage support

seancorfield opened the pull request:
Adds `nvd.task` namespace with `check` task that invokes the `-main` function.

Adds minimal documentation.

vemv commented:
Cheers! I'll iterate on this one keeping attribution. In particular we need to extend `.github/integration_test.sh` - feel free to give it a shot but certainly not required

seancorfield commented:
Something that may also be worth adding to the README about this is that using `nvd-clojure` as a "tool" like this ensures classpath isolation since `clojure -T...` ignores the project `deps.edn` context.

I'll take a look at the integration test stuff next week (I'm on vacation all week and will probably spend a day or two doing OSS stuff).

vemv commented:
Yes SGTM, that section of the README predates `-T` so I don't mind refactoring it to keep it idiomatic.

I'll open a PR against your branch I think. Either way happy holiday 🍻

seancorfield commented:
You probably want to edit the description of this project on GH since it still indicates it is Leiningen only (it was listed in this week's Deref!).

seancorfield commented:
> We got to update it finally! Thanks.

The repo still has the description that is Leiningen-specific:

> National Vulnerability Database dependency-checker for deps.edn and plug-in for Leiningen

vemv commented:
One can observe that colloquially, a project is either "deps.edn" (or maybe "tools.deps") or "Leiningen".

I haven't heard people saying "my project is managed by the Clojure CLI", seems like bit of a mouthful (and maybe not always correct: one can use tools.deps from programs aside from via CLI).

Simplifying things, the description could simply say `National Vulnerability Database dependency checker for Clojure projects`. Which is particularly accurate: the usage we intend to foster is a simple program that accepts a classpath as an argument (this particularly discards "plugins"). That is compatible with all sorts of projects, e.g. also those managed by Boot, Maven, Gradle or what have you.

vemv commented on .github/integration_test.sh:
```diff
@@ -1,21 +1,27 @@
 
-cd .. || exit 1
-cd example || exit 1
+if ! clojure -Ttools install nvd-clojure/nvd-clojure '{:mvn/version "RELEASE"}' :as nvd; then
```
Added the Tools way to the integration test, it's passing 👍

vemv commented on .github/integration_test.sh:
```diff
@@ -87,7 +93,28 @@ if ! grep --silent "$SUCCESS_REGEX" example-lein-output; then
+# cd to $HOME, to demonstrate that the Tool does not depend on a deps.edn file:
+cd || exit 1
+
+if clojure -Tnvd nvd.task/check :classpath '"'$example_classpath'"' > example-lein-output; then
```
I had to add two tweaks over what was hinted over the readme:

* had to qualify `nvd.task/check`, the CLI wouldn't be happy otherwise
* changed `io.github` Maven prefix for `nvd-clojure`, which matches our project.clj structure

seancorfield replied:
Re: qualification -- that indicates you don't have the correct `:tools/usage` in `deps.edn` to specify the default namespace. Ah, most likely because you're installing from the Clojars version and it doesn't have `deps.edn` in the JAR?

Normally, CLI tool installation is done from GitHub, not Clojars/Maven, and you use `io.github.rm-hull/nvd-clojure` as the lib name (group/artifact => GitHub URL) and just specify a tag `{:git/tag "v1.2.3"}`

vemv commented on README.md:
```diff
@@ -28,7 +28,19 @@ dependencies and passes them to a library called [Dependency-Check](https://gith
-To install globally, add `nvd-clojure/nvd-clojure {:mvn/version "1.8.0"}`
+To install in a given project, you can add `nvd-clojure/nvd-clojure {:mvn/version "1.8.0"}` to your deps.edn.
+
+If you have CLI version 1.10.3.933 or later, you can also install `nvd-clojure` as a "tool":
```
For a brief period, Tools usage will be an 'also' thing. Shortly thereafter we can remove all references to risky/unaccurate ways of running nvd-clojure (it was about time!).

vemv commented on src/nvd/task.clj:
```diff
@@ -0,0 +1,31 @@
+
+(defn check
+  "Arguments: `:config-filename` (optional), `:classpath` (required)."
+  [{:keys [config-filename classpath]}]
```
renamed config -> config-filename, keeping the terminology consistent, relative to `-main`

vemv commented on src/nvd/task/check.clj:
```diff
@@ -126,6 +126,14 @@
+                                    ;; source paths such as `src`, while are part of the classpath,
+                                    ;; won't be meaningfully analyzed by dependency-check-core.
+                                    ;; Keeping only .jars facilitates various usage patterns.
+                                    (s/ends-with? s ".jar")))
```
Fixes something @seancorfield pinged me about some time ago (and was failing the integration test anyway)

vemv reviewed (commented):
Rebased/amended incorporating various feedback directly

vemv reviewed (approved):
LGTM (but I'm biased 😇)

@seancorfield , let me know if these tweaks look reasonable. Particularly interested in whether the integration test is both accurate and representative of actual 'UX'.

Feel free to push things directly, I won't push further

vemv commented:
As for the build failures:

> -T is no longer supported, use -A with repl, -M for main, or -X for exec

Might be that we're using an older CLI?

> Error: Error: Command failed: git cat-file -p f20c2c031ac6b343325a437549d2432a153e6b1d
> fatal: Not a valid object name f20c2c031ac6b343325a437549d2432a153e6b1d

Probably Coveralls isn't fork-safe, feel free to ignore

seancorfield commented:
If you're getting "-T is no longer supported" then, yes, your Clojure CLI version for CI is too old.

vemv commented:
Alright, sounding good! Feel free to update the CLI in our `.github/` and to ensure it works without qualification.

Relatedly, obviously an essential part of the integration test is that Tools runs the github repo's version as of $commit, not one fetched over HTTP.

I believe that is currently achieved by the following means:

* first one `lein install`s a Maven artifact
* then, one consumes said local artifact here: `clojure -Ttools install nvd-clojure/nvd-clojure '{:mvn/version "RELEASE"}'`

I'm mention this in case it affects anything you might want to change.

seancorfield commented:
I've updated the CLI version (in two places) and updated how the tool is installed (local/root in CI, git/tag in the README), and unqualified the function in `clojure -Tnvd check` since that should work as-is with the updated install command.

vemv commented on README.md:
````diff
@@ -33,7 +33,7 @@ To install in a given project, you can add `nvd-clojure/nvd-clojure {:mvn/versio
 
 ```bash
-clojure -Ttools install nvd-clojure/nvd-clojure '{:mvn/version "RELEASE"}' :as nvd
+clojure -Ttools install io.github.rm-hull/nvd-clojure '{:git/tag "1.8.0"}' :as nvd
````
I'd prefer this was a Maven version, since it's the one thing that we always release (today I released 1.8.0 and forgot to tag for example*)

*will add it to the Makefile that performs deploys. Then again, people can deploy things by hand.

vemv commented on .github/integration_test.sh:
```diff
@@ -17,7 +17,7 @@ if ! lein with-profile -user,-dev,+ci install; then
 fi
 
-if ! clojure -Ttools install nvd-clojure/nvd-clojure '{:mvn/version "RELEASE"}' :as nvd; then
+if ! clojure -Ttools install io.github.rm-hull/nvd-clojure '{:local/root "."}' :as nvd; then
```
What does `io.github.rm-hull` exactly do/add? 

I tend to run `.github/integration_test.sh` locally often. Having `io.github.rm-hull` there invites one to think one might be running something else (as much as I can read `root .`)

vemv reviewed (commented):
CI is failing with `Error building classpath. Bad coordinate for library , expected map: nil`

Another thing, if the unqualifed name requires `mvn` not to be used, I'm not sure I'd be comfortable with that. I don't want to promote a specific preference just because it happens to be shorter due to some tool's specific constraints.

Users concerned with succinctness can trivially add a bash `alias` (`clojure -Tfoo` is a mouthful anyway?)

seancorfield commented:
> CI is failing with `Error building classpath. Bad coordinate for library , expected map: nil`

I missed that the `lein install` line is in the `plugin` folder. The tools install should happen in the root folder. I'll swap that above the `cd plugin` line.

> Another thing, if the unqualifed name requires `mvn` not to be used, I'm not sure I'd be comfortable with that.

It's more common for CLI tools to be installed from source, directly from GitHub, and thus `:tools/usage` in `deps.edn` is respected and you can use functions in the default namespace without qualification. This is standard practice with the CLI.

If you use `:mvn/version`, you lose the `:tools/usage` functionality and you have to use fully-qualified function names.

> What does `io.github.rm-hull` exactly do/add?

I didn't realize you were unfamiliar with how git deps are specified in recent CLI versions, sorry. The standard practice is to specify tools (and any other git deps) as `io.github.<org>/<repo>` so you can use `:git/tag` and/or `:git/sha` without needing to specify `:git/url` -- the GitHub URL is automatically derived from the lib name: `https://github.com/<org>/<repo>`

Since you don't seem very comfortable with installing the tool from GitHub, I've reverted to `:mvn/version` and added back the namespace qualifier on the function name in `clojure -Tnvd`.

The big benefit of the git deps approach is that user can install the tool from a branch (or an arbitrary SHA) in order to test/use features that have not yet been released. I have been testing the `nvd` tool via the SHA of this PR, e.g.,

```
clojure -Ttools install io.github.rm-hull/nvd-clojure '{:git/sha "afae9e7c1f26e67321b0d675dfe3308be49d89c5"}' :as nvd
```

vemv commented:
Thanks a lot for your time and incorporating the feedback!

I've released these as 1.9.0.

The integration test is failing with `Error building classpath. Bad coordinate for library , expected map: nil`. I cannot reproduce it locally (integration_test.sh passes for me).

https://github.com/rm-hull/nvd-clojure/runs/4258339536?check_suite_focus=true

Any ideas welcome.

seancorfield commented:
I submitted a PR to update `integration_test.sh` with two CLI commands that might help me debug things...

seancorfield commented:
Tried several things. Still puzzled. Posted in tools-deps on Slack for more input...

seancorfield commented:
Looks like it's a bug in the CLI script itself. I've updated that patch-1 PR to the minimum that I think will work but it's a pure hack based on internal knowledge of the CLI stuff :)

seancorfield commented:
Tracked it down to a bug in the `setup-clojure` script -- https://github.com/DeLaGuardo/setup-clojure/issues/34 -- so there's nothing to be done until that is fixed. I can submit a clean PR that will add a workaround to the `integration_test.sh` script if you'd like?
